From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stefan Kristiansson <stefank@nvidia.com>
Date: Mon, 8 Apr 2024 12:50:45 +0300
Subject: [PATCH 246/254] NVIDIA: SAUCE: firmware: tegra: bpmp: add die
 specific debugfs nodes

In a multidie environment, bpmp debugfs nodes have to be created in
separate directories for each die.
Create directories with the die number as a suffix.
Create a symlink without suffix to die zero directory, for backwards compatibility.

http://nvbugs/4466065

Change-Id: I59a8f27fcb4d18b71cee96bc51701d3f221d6fc8
Signed-off-by: Stefan Kristiansson <stefank@nvidia.com>
(cherry picked from commit 5d7f878f74aa263a7ebfb0695a844fea845e5f9d)
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3180933
Reviewed-by: Rakesh Babu Bodla <rbodla@nvidia.com>
Tested-by: Rakesh Babu Bodla <rbodla@nvidia.com>
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
(cherry picked from commit 579059e1029d4c11fd4ec4aed2907f1ba4da8335)

diff --git a/drivers/firmware/tegra/bpmp-debugfs.c b/drivers/firmware/tegra/bpmp-debugfs.c
index 4221fed70ad48..5362e3ce51929 100644
--- a/drivers/firmware/tegra/bpmp-debugfs.c
+++ b/drivers/firmware/tegra/bpmp-debugfs.c
@@ -772,6 +772,8 @@ static int bpmp_populate_debugfs_shmem(struct tegra_bpmp *bpmp)
 int tegra_bpmp_init_debugfs(struct tegra_bpmp *bpmp)
 {
 	struct dentry *root;
+	struct dentry *symlink = NULL;
+	char tmp[10];
 	bool inband;
 	int err;
 
@@ -780,10 +782,19 @@ int tegra_bpmp_init_debugfs(struct tegra_bpmp *bpmp)
 	if (!inband && !tegra_bpmp_mrq_is_supported(bpmp, MRQ_DEBUGFS))
 		return 0;
 
-	root = debugfs_create_dir("bpmp", NULL);
+	if (dev_to_node(bpmp->dev) == NUMA_NO_NODE) {
+		root = debugfs_create_dir("bpmp", NULL);
+	} else {
+		snprintf(tmp, sizeof(tmp), "bpmp%d", dev_to_node(bpmp->dev));
+		root = debugfs_create_dir(tmp, NULL);
+	}
+
 	if (IS_ERR(root))
 		return -ENOMEM;
 
+	if (dev_to_node(bpmp->dev) == 0)
+		symlink = debugfs_create_symlink("bpmp", NULL, "bpmp0");
+
 	bpmp->debugfs_mirror = debugfs_create_dir("debug", root);
 	if (IS_ERR(bpmp->debugfs_mirror)) {
 		err = -ENOMEM;
@@ -797,8 +808,10 @@ int tegra_bpmp_init_debugfs(struct tegra_bpmp *bpmp)
 		err = bpmp_populate_debugfs_shmem(bpmp);
 
 out:
-	if (err < 0)
+	if (err < 0) {
+		debugfs_remove(symlink);
 		debugfs_remove_recursive(root);
+	}
 
 	return err;
 }
-- 
2.43.0

